scipy/sparse/_coo.py

```diff
@@ -499,9 +517,10 @@ def _setdiag(self, values, k):
             new_data[:] = values
 
         # Update the internal structure.
-        self.row = np.concatenate((self.row[keep], new_row))
-        self.col = np.concatenate((self.col[keep], new_col))
-        self.data = np.concatenate((self.data[keep], new_data))
+        xp, _ = get_namespace(self.row, self.col)
+        self.row = xp.asarray(np.concatenate((self.row[xp.asarray(keep)], new_row)))
```

Comment:
If `self.row` is an `array_api` array on cupy, that means `np.concatenate` will move the data into CPU and then `xp.asarray` will bring it back to the GPU.